### ISSUE TITLE
AfterAllCallback should not be invoked when exception is thrown in constructor or static block

### ISSUE THREAD

**Issue by beluha** [junit-team/junit5#1460]
## Overview

The `afterAll(ExtensionContext)` callback method from the `AfterAllCallback` interface is invoked when an exception is thrown in a constructor static initialization block when using `@TestInstance(Lifecycle.PER_CLASS)`.

However, no other callback methods are invoked. `@BeforeAll` and `@AfterAll` methods are also not executed.

### Test Class

```java
@TestInstance(Lifecycle.PER_CLASS)
@ExtendWith(LoggerExtension.class)
class Example {

	Example() {
		System.out.println("Constructor");
		throw new RuntimeException("Constructor exception");
	}

	@BeforeAll
	void beforeAll() {
		System.out.println("@BeforeAll");
	}

	@AfterAll
	void afterAll() {
		System.out.println("@AfterAll");
	}

	@BeforeEach
	void beforeEach() {
		System.out.println("\t@BeforeEach");
	}

	@AfterEach
	void afterEach() {
		System.out.println("\t@AfterEach");
	}

	@org.junit.jupiter.api.Test
	void exampleTest() {
		System.out.println("\t\t\t@Test");
	}

}
```

### LoggerExtenstion

```java
public class LoggerExtension implements BeforeEachCallback, AfterEachCallback, BeforeAllCallback, AfterAllCallback,
		BeforeTestExecutionCallback, AfterTestExecutionCallback {

	@Override
	public void beforeEach(ExtensionContext context) throws Exception {
		System.out.println("\tBeforeEachCallback");
	}

	@Override
	public void afterEach(ExtensionContext context) throws Exception {
		System.out.println("\tAfterEachCallback");
	}

	@Override
	public void beforeTestExecution(ExtensionContext context) throws Exception {
		System.out.println("\t\tBeforeTestExecutionCallback");
	}

	@Override
	public void afterTestExecution(ExtensionContext context) throws Exception {
		System.out.println("\t\tAfterTestExecutionCallback");
	}

	@Override
	public void beforeAll(ExtensionContext context) throws Exception {
		System.out.println("BeforeAllCallback");
	}

	@Override
	public void afterAll(ExtensionContext context) throws Exception {
		System.out.println("AfterAllCallback");
	}

}
```

## Actual Behavior

```
Constructor
AfterAllCallback
```

The behavior is the same if `@BeforeAll` and `@AfterAll` are declared as `static`.

## Expected Behavior

An `AfterAllCallback` should not be invoked unless a `BeforeAllCallback` is invoked, per discussion in [Gitter](https://gitter.im/junit-team/junit5).

## Related Issues

- #359

## Deliverables

- [ ] Ensure that `AfterAllCallback`s are **not** invoked if an exception is thrown by the test class constructor.

**Comment by sbrannen** [comment]
Behavior when no exception is thrown in the constructor:

```
Constructor
BeforeAllCallback
@BeforeAll
	BeforeEachCallback
	@BeforeEach
		BeforeTestCallback
			@Test
		AfterTestCallback
	@AfterEach
	AfterEachCallback
@AfterAll
AfterAllCallback
```

**Comment by sbrannen** [comment]
FYI: I modified the output text to include indentation via tabs.

**Comment by sbrannen** [comment]
I can confirm that the following is the output when an exception is thrown in the constructor:

```
Constructor
AfterAllCallback
```

**Comment by sbrannen** [comment]
_in progress_

**Comment by sbrannen** [comment]
This issue is closely related to #359.

**Comment by sbrannen** [comment]
I already have a fix in place based on the work I did in #359.

So I basically just need to put the appropriate test in place. 😉

**Comment by sbrannen** [comment]
Hmmm.... looks like I spoke too soon: `ExceptionHandlingTests.exceptionInAfterAllCallbackDoesNotHideFailureWhenTestInstancePerClassIsUsed()` now fails with my proposed _fix_.

**Comment by gaganis** [comment]
In attempt to convey our discussion on gitter, I think we discussed 3 different topics. All the case below are referring to Classes annotated with the `PER_CLASS` instantiation method.

- a) Why jupiter would not run instance methods annotated with `@BeforeAll` or `@AfterAll` when we have a failing constructor. The conclusion for this was that it is _impossible_ for these methods to run since no instance can be created.
- b) Why jupiter would not run static methods annotated with `@BeforeAll` or `@AfterAll` when we have a failing constructor. Technically there is no restriction why these would not run so it is a matter of design what would the expected behavior here. 
On the one hand have a `PER_CLASS` class failing on the constructor suggests a major problem that warrants aborting everything. Also why run BeforeAll of AfterAll when we have run none tests? 
On the other hand this seems inconsistent with what we do for classes that are instantiated per method where a failing constructor does not affect the execution of methods annotated with `@BeforeAll` or `@AfterAll`.  So I personally consider this as unexpected behavior.
- c) Why jupiter would run AfterAllCallback while not running everything else. By the current title of this issue this point is the one handled by this issue. 

Also my understanding from the conversation on gitter was that @beluha raised the issue because she also expected that @BeforeAll and @AfterAll callbacks should be running. @beluha Could you please clarify what you expected? Thanks

**Comment by sbrannen** [comment]
> - a) Why Jupiter would not run instance methods annotated with `@BeforeAll` or `@AfterAll` when we have a failing constructor. The conclusion for this was that it is _impossible_ for these methods to run since no instance can be created.

That's correct: we cannot achieve the literally impossible. 😉

**Comment by sbrannen** [comment]
> - c) Why Jupiter would run `AfterAllCallback` while not running everything else.
  - By the current title of this issue this point is the one handled by this issue. 

That's correct: as a general rule of thumb, JUnit Jupiter never runs AfterX methods or callbacks unless it ran (or attempted to run) BeforeX methods or callbacks.

The framework has been consistent in this regard, and whenever notice anything to the contrary... we fix it (like in this issue).

**Comment by sbrannen** [comment]
> Also my understanding from the conversation on Gitter was that @beluha raised the issue because she also expected that `@BeforeAll` and `@AfterAll` callbacks should be running.

I'm not sure if you are referring to the callbacks (extensions) or user methods here.

In any case, I understand her desire differently.

So, yes, @beluha, could you please clarify what you expected?

**Comment by sbrannen** [comment]
@gaganis, thanks for taking the time to document the discussion on Gitter!

Regarding point _B_, that's an interesting debate but beyond the scope of _this_ GitHub issue.

Would you mind opening a separate issue to discuss that on its own?

**Comment by sbrannen** [comment]
FYI: the current work on this issue can be viewed in the following feature branch.

https://github.com/junit-team/junit5/commits/issues/1460-invoke-AfterAllCallback-only-if-BeforeAllCallbacks-invoked

**Comment by gaganis** [comment]
> I'm not sure if you are referring to the callbacks (extensions) or user methods here.

Sorry there is a typo. Please replace _callbacks_ -> _methods_



> Regarding point b), that's an interesting debate but beyond the scope of this GitHub issue.
> 
> Would you mind opening a separate issue to discuss that on its own?

Sure, will do ;)

**Comment by beluha** [comment]
For me BeforeAllCallback and AfterAllCallback should not be invoked, if an  exception is thrown in constructor.

**Comment by sbrannen** [comment]
This bug has been fixed in 7ada7e823a2afcece4fc7af715a7ded6b17f62d8 and further polished in 9693d53891219547963fb0afc4b82979c03732fa and 5d1e6ef203abe68e2a601fa7c9283d630e698e4b.

**Comment by sbrannen** [comment]
@beluha, thanks for the clarification!

The aforementioned fix makes it work like you expect.
